middle_end/closure/closure.ml

```diff
@@ -697,15 +697,39 @@ let rec substitute loc ((backend, fpc) as st) sb rn ulam =
   | Uunreachable ->
       Uunreachable
 
-(* Perform an inline expansion *)
+(* Perform an inline expansion:
 
+   If [f p = body], substitute [f a] by [let p = a in body].
+
+   Under certain conditions, further simplifications are possible (we use the
+   terminology of [Semantics_of_primitives], applied to terms of the Clambda
+   language):
+
+   - [f a] is equivalent to [body[a/p]] if [a] has no effects and no coeffects.
+     However, we only want to do this rewriting if [body[a/p]] does not increase
+     the size of [body]. Since this is hard to decide in general, as an
+     approximation, only consider the case when [a] is an immutable variable or
+     a constant.
+
+   - [f a] is equivalent to [body] if [p] does not occur in [body] and [a] has
+     only generative effects. As an approximation we only do this when [a] is
+     either a variable or a constant or a closure allocation.
+
+   - In general [f a] is equivalent to [a; body] if [p] does not occur in [body].
+*)
+
+let mutable_ids = Hashtbl.create 17
+
+(* Approximates "no effects and no coeffects" *)
 let is_simple_argument = function
```

Comment:
Could this function be maybe renamed `is_substituable`?

Reply:
Done.